Optimize BitUtils::forEachBit

kgpai opened the pull request:
Currently large SelectivityVectors, say with 1k+ rows (e.g element vector of an array vector) are slower when they have any one row not selected vs all rows selected. This is because even if one row is not selected we default to slowest way of iterating through the vector. Typically this results in the iteration being 100% slower. With this fix to forEachBit, applyToSelected() runs faster but also more consistent. This will also gracefully degrade to slowest way of iterating through vector based on number of unset bits. 


```
----------------------------------------------------------------------------
BM_isAllSelectedTrue(1000)                                 315.71us    3.17K
BM_isAllSelectedFalseLastBit(1000)                58.82%   536.78us    1.86K
BM_isAllSelectedFalseFirstBit(1000)               58.79%   537.06us    1.86K
----------------------------------------------------------------------------
============================================================================
```
Heres new :
```
----------------------------------------------------------------------------
BM_isAllSelectedTrue(1000)                                 316.38us    3.16K
BM_isAllSelectedFalseLastBit(1000)                92.80%   340.91us    2.93K
BM_isAllSelectedFalseFirstBit(1000)               89.65%   352.89us    2.83K
----------------------------------------------------------------------------
============================================================================
```

**How it works:**
* We take one word at a time
* Check if all bits are set, if true : Then run func() over set bits using for loop . (fastest)
* Else if any one bit is unset, use while loop to individually run func() over selected bits. (slowest)
* ForEachWord, takes a partial word function, and a full word function which implement the above .

mbasmanova commented on velox/vector/SelectivityVector.h:
```diff
@@ -364,12 +364,46 @@ class SelectivityIterator {
     }
   } else {
-    bits::forEachSetBit(bits_.data(), begin_, end_, func);
+    bits::forEachWord(
```
Is bits::forEachWord faster then bits::forEachSetBit ? I wonder why would that be the case given that forEachSetBit is implemented using forEachWord .

kgpai replied:
I have added a brief explanation in the description above, let me know if that doesnt help.

mbasmanova replied:
I wonder if the same improvement can be achieved by optimizing forEachBit, e.g. changing the following code to add fast path is word is all-set or all-unset.

```
     [isSet, bits, func](int32_t idx) {
        auto word = (isSet ? bits[idx] : ~bits[idx]);
        if (!word) {
          return;
        }
        while (word) {
          func(idx * 64 + __builtin_ctzll(word));
          word &= word - 1;
        }
      }
```

kgpai replied:
To make sure we are on same page, I think you are suggesting we move this to logic to forEachBit. I think that will make forEachBit faster - though in isAllSelected, we will still pay cost of checking if all bits are selected. Would it make sense to add this same logic in both places ? Nevertheless let me move this logic there and see what performance change I observe.

kgpai replied:
@mbasmanova Made the changes to forEachBit and I am seeing similar numbers:

```
Current implementation:
============================================================================
/Users/kpai/src/Velox/velox/vector/benchmarks/SelectivityVectorBenchmark.cpprelative  time/iter  iters/s
============================================================================
BM_isAllSelectedTrue(100000)                                24.02ms    41.64
BM_isAllSelectedFalseLastBit(100000)              49.76%    48.27ms    20.72
BM_isAllSelectedFalseFirstBit(100000)             49.89%    48.14ms    20.77
----------------------------------------------------------------------------
============================================================================

New implementation:
============================================================================
/Users/kpai/src/Velox/velox/vector/benchmarks/SelectivityVectorBenchmark.cpprelative  time/iter  iters/s
============================================================================
BM_isAllSelectedTrue(100000)                                26.20ms    38.16
BM_isAllSelectedFalseLastBit(100000)              96.88%    27.05ms    36.97
BM_isAllSelectedFalseFirstBit(100000)            102.44%    25.58ms    39.10
----------------------------------------------------------------------------
============================================================================
```

mbasmanova commented:
> Made the changes to forEachBit and I am seeing similar numbers:

@kgpai Thank you for checking. Sounds like the optimization can be re-implemented by changing the implementation of forEachBit where it would benefit more code paths and would allow to preserve somewhat short and simple implementation for applyToSelected and testSelected.

kgpai commented:
Update : Moved changes to forEachBit.

mbasmanova commented on velox/common/base/BitUtil.h:
```diff
@@ -345,19 +346,35 @@ void forEachBit(
-        while (word) {
-          func(idx * 64 + __builtin_ctzll(word));
-          word &= word - 1;
+        if (kAllSet == bits[idx]) {
```
Shouldn't this comparison use "word"? How would this work when the function is called with isSet == false? Looks like we might be missing some unit tests. Let's add.

mbasmanova commented on velox/common/base/BitUtil.h:
```diff
@@ -345,19 +346,35 @@ void forEachBit(
-        while (word) {
-          func(idx * 64 + __builtin_ctzll(word));
-          word &= word - 1;
+        if ((kAllSet & mask) == (word & mask)) {
```
Is this optimization important? Do you see a difference in the benchmark results?

The logic seems to rely on mask representing contiguous bits. If so, perhaps, it would be clearer to use __lzcnt64 intrinsic (count number of leading zeros) to calculate "last". __builtin_popcountll calculates number of bits set and gives an impression that random bits might be set. 

It would be nice to add a comment to clarify that mask is expected to represent contiguous set of bits.

mbasmanova commented on velox/common/base/BitUtil.h:
```diff
@@ -345,19 +346,35 @@ void forEachBit(
-          word &= word - 1;
+        if ((kAllSet & mask) == (word & mask)) {
+          const size_t start = idx * 64 + __builtin_ctzll(word);
+          const size_t last = start + __builtin_popcountll(word & mask);
```
& mask is redundant here; let's remove

mbasmanova reviewed (commented):
@kgpai With this optimization, do you no longer see the overhead of calling isAllSelected ? Perhaps, add a benchmark for isAllSelected to confirm that it became a lot faster.

kgpai commented:
- Simplified forEachBit so we only use fast case when all bits are set
- Added Test case 
- Added benchmark , benchmark results are in description.

mbasmanova commented on velox/common/base/tests/BitUtilTest.cpp:
```diff
@@ -0,0 +1,65 @@
+namespace velox {
+namespace test {
+
+TEST(BitUtilTest, forEachBit) {
```
There is an existing test for bits manipulating functions in velox/common/tests/BitUtilTest.cpp . It would be nice to move this test there.

kgpai replied:
You are right, I did not see any tests under velox/common/base, where bitUtilTest is and presumed there are no tests .

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+
+  auto numEntries = 100;
+  uint32_t size = bits::nwords(numEntries);
+  std::vector<uint64_t> vec(size);
```
do not abbreviate; some better names here could be data and rawData.

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
   EXPECT_EQ(totalBits - 3, expectedIndex);
+
+  auto numEntries = 100;
+  uint32_t size = bits::nwords(numEntries);
```
nit: size is used only once, hence, consider inlining: `data(bits::nwords(totalBits))`

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
   forEachUnsetBit(data, 2, totalBits - 3, calledOnEveryOther);
   EXPECT_EQ(totalBits - 3, expectedIndex);
+
+  auto numEntries = 100;
```
Entries is a strange term to use here. Let's use bits. Perhaps, just reuse totalBits variable.

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+  std::vector<uint64_t> vec(size);
+  uint64_t* vecPtr = vec.data();
+
+  // Test out iterating over all entries when true
```
Comments should be full sentences, e.g. end with a period. Please fix throughout.

"all entries when true" sounds strange; how about "Test all bits set." and "Test all bits unset."

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+  // Test out iterating over all entries when true
+  bits::fillBits(vecPtr, 0, numEntries, true);
+  int count = 0;
+  bits::forEachBit(vecPtr, 0, numEntries, true, [&](auto row) {
```
nit: consider introducing temp variable for the lambda to avoid copy-paste

```
auto countEach = [&](auto row) {
    ASSERT_EQ(row, count);
    count++;
  };

bits::forEachSetBit(rawData, 0, totalBits, countEach);
ASSERT_EQ(totalBits, count);

...
bits::forEachUnsetBit(rawData, 0, totalBits, countEach);
ASSERT_EQ(totalBits, count);
```

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+  });
+  ASSERT_EQ(numEntries, count);
+
+  // Unset one entry when everything is set.
```
nit: perhaps, "Test all but one bit set." and "Test all but one bit unset."

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
 
+void BM_forEachBit(
+    uint32_t iterations,
+    size_t numEntries,
```
numBits ?

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+    std::optional<size_t> bitPos = std::nullopt) {
+  folly::BenchmarkSuspender suspender;
+  uint32_t size = bits::nwords(numEntries);
+  std::vector<uint64_t> vec(size);
```
data and rawData for clarity

```
std::vector<uint64_t> data(bits::nwords(numBits));
auto* rawData = data.data();
```

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+void BM_forEachBit(
+    uint32_t iterations,
+    size_t numEntries,
+    bool allSelected,
```
this flag is not needed; it is the same as bitPos.has_value()

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+    uint32_t iterations,
+    size_t numEntries,
+    bool allSelected,
+    std::optional<size_t> bitPos = std::nullopt) {
```
perhaps, clarify that this is the bit to unset, e.g. bitToClear or unsetBit

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+  suspender.dismiss();
+
+  for (uint32_t i = 0; i < iterations; ++i) {
+    bits::forEachBit(vecPtr, 0, numEntries, true, [&](auto row) {
```
[&] -> [] as lambda doesn't use any non-local variables

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+const static auto kNumRuns = 1000;
+}
+void BM_ForEachBitAllTrue(uint32_t iterations, size_t numEntries) {
+  for (int i = 0; i < kNumRuns; i++) {
```
Why do you need both iterations and kNumRuns?

kgpai replied:
iterations isnt big enough, we typically get values in micro or nano seconds, kNumRuns is just to amplify those runs.

mbasmanova replied:
Why not just increase "iterations"?

funrollloops replied:
The absolute time per iteration does matter for sanity checking and for comparing to other benchmark results. For example, the cost per row here should affect the cost per row of almost every other function in the codebase.

For example: your current benchmark results are ~300us or ~3k-2k iters/sec. This is for 1000 x 10k blocks, so we are talking about 30-20B rows/sec, or 0.03ns/row. It's useful to have this as a datapoint when looking at the e.g. benchmarks of the simple UDF interface.

You can use the `_MULTI` family of functions to change how the iterations and time/iteration are displayed; IMO we should try to standardize on an iteration being a row in benchmark output.
https://github.com/facebook/folly/blob/main/folly/Benchmark.h#L392

mbasmanova commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,55 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+namespace {
+const static auto kNumRuns = 1000;
+}
+void BM_ForEachBitAllTrue(uint32_t iterations, size_t numEntries) {
```
numEntries -> numBits

mbasmanova reviewed (commented):
Looks good to me. Some comments for the test and benchmark below.

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+
+  totalBits = 100;
+  std::vector<uint64_t> bits(bits::nwords(totalBits));
+  uint64_t* rawData = bits.data();
```
naming: rawBits

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+  // Test all bits set.
+  bits::fillBits(rawData, 0, totalBits, true);
+  int count = 0;
+  auto countEachFn = [&](auto row) {
```
do not abbreviate: countEach

mbasmanova commented on velox/common/tests/BitUtilTest.cpp:
```diff
@@ -416,6 +416,43 @@ TEST_F(BitUtilTest, forEachBit) {
+  bits::fillBits(rawData, 0, totalBits, true);
+  bits::setBit(rawData, 50, false);
+  count = 0;
+  auto incrementCountFn = [&](auto row) { count++; };
```
do not abbreviate: countSelected, maybe? to match countEach above

mbasmanova reviewed (approved):
@kgpai Looks good to me.

funrollloops commented on velox/common/base/BitUtil.h:
```diff
@@ -337,6 +337,7 @@ void forEachBit(
     int32_t end,
     bool isSet,
     Callable func) {
+  static const uint64_t kAllSet = -1ULL;
```
nit: constexpr

funrollloops commented on velox/common/base/benchmarks/BitUtilBenchmark.cpp:
```diff
@@ -141,6 +141,54 @@ BENCHMARK_RELATIVE(BM_scatterBits, n) {
+    uint32_t iterations,
+    size_t numBits,
+    std::optional<size_t> bitToClear = std::nullopt) {
+  folly::BenchmarkSuspender suspender;
```
Can you pass kNumRuns * iterations as `iterations` here to avoid re-doing the setup and teardown?

funrollloops commented on velox/common/base/BitUtil.h:
```diff
@@ -355,9 +356,17 @@ void forEachBit(
         if (!word) {
```
Do we really need this branch? the `while(word)` below will already exit immediately if word is zero.

funrollloops commented on velox/common/base/BitUtil.h:
```diff
@@ -355,9 +356,17 @@ void forEachBit(
+          const size_t start = idx * 64;
+          const size_t end = (idx + 1) * 64;
+          for (size_t row = start; row < end; ++row) {
+            func(row);
```
In a microbenchmark where `func` is small and therefore doesn't affect inlining, I would definitely expect to see a win from this change. However now `func` is invoked twice, which could cause code size to increase and affect inlining decisions. Have you run other function benchmarks to check the impact of this change?

kgpai replied:
Thats a great point ; Looking at clang documentation(https://clang.llvm.org/docs/analyzer/developer-docs/IPA.html#deciding-when-to-inline)  though It looks like function length isnt the primary factor that determines inlining but more other factors (variadic functions, fn definition not avilable etc..). I would reckon in most cases it should inline, however I will try and run some tests with larger functions to determine this.

funrollloops replied:
Discussed offline: other benchmarks in the codebase are not negatively impacted, so we're fine here.
